foucaultview.cpp

```diff
@@ -152,6 +169,328 @@ QVector<QPoint> scaleProfile(QPolygonF points, int width,
 
     return results;
 }
+
+
+QImage foucaultView::generateOpticalTestImage(OpticalTestType type, wavefront* wf, const OpticalTestSettings& s)
+{
+    if (!wf || wf->data.cols == 0) return QImage();
+
+    // 1. Setup Constants
+    double pad = 1.1;
+    int size = (int(wf->data.cols * pad) / 2) * 2;
+    double actualPad = (double)size / wf->data.cols;
+    double moving_constant = (s.movingSource) ? 1.0 : 2.0;
+
+    mirrorDlg *md = mirrorDlg::get_Instance();
+    double unitMultiplyer = s.useMM ? 1.0 : 25.4;
+    double coc_offset_mm = s.rocOffset * unitMultiplyer;
+
+    // Physics geometry
+    double r2 = (md->diameter / 2.0) * (md->diameter / 2.0);
+    double b = md->roc + coc_offset_mm;
+    double pv = (sqrt(r2 + (md->roc * md->roc)) - (sqrt(r2 + b * b) - coc_offset_mm)) / (md->lambda * 1.E-6);
+    double z3 = pv / moving_constant;
+    double effectiveZ3 = (type == OpticalTestType::Ronchi) ? (s.ronchiX * z3) : z3;
+
+    // 2. Wavefront Prep
+    std::vector<double> originalZerns = wf->InputZerns;
+    std::vector<double> tempZerns = originalZerns;
+    tempZerns[3] -= 3 * tempZerns[8];
+    wf->InputZerns = tempZerns;
+
+    SimulationsView *sv = SimulationsView::getInstance(0);
+    sv->setSurface(wf);
+    cv::Mat surf_fft = sv->computeStarTest(s.heightMultiply * sv->nulledSurface(effectiveZ3), size, actualPad, true);
+    wf->InputZerns = originalZerns; // Restore state immediately
+
+    // 3. Mask Generation
+    cv::Mat mask = cv::Mat::zeros(size, size, CV_64FC1);
+    cv::Mat sourceSlit = cv::Mat::zeros(size, size, CV_64FC1);
+    int hx = (size - 1) / 2 + s.lateralOffset;
+    double pixwidth = s.outputLambda * 1.E-6 * (0.5 * md->roc / md->diameter) * 2. / (25.4 * actualPad);
+
+    if (type == OpticalTestType::Ronchi) {
+        double lpi_val = s.lpi * (s.useMM ? 25.4 : 1.0);
+        int ppl = std::max(1, (int)((0.5 / lpi_val) / pixwidth));
+        int start = (size / 2) - (ppl / 2);
+        bool even = ((start / ppl) % 2 == 0) ^ s.clearCenter;
```

Comment:
```suggestion
        bool even = ((start / ppl) % 2 == 0) != s.clearCenter;
```

^ is a bitwise operator.

Reply:
Yes I think in this case ^ is correct.  This was code generated by AI  a clever trick to test the int result being 0 or 1 along with the test for a clear center setting.